ipfs/embedded.go

```diff
@@ -0,0 +1,154 @@
+package ipfs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	ipfscfg "github.com/ipfs/go-ipfs-config"
+	"github.com/ipfs/go-ipfs/commands"
+	"github.com/ipfs/go-ipfs/core"
+	"github.com/ipfs/go-ipfs/core/coreapi"
+	"github.com/ipfs/go-ipfs/core/corehttp"
+	"github.com/ipfs/go-ipfs/core/node/libp2p"
+	"github.com/ipfs/go-ipfs/repo"
+	"github.com/ipfs/go-ipfs/repo/fsrepo"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+
+	"github.com/lazyledger/lazyledger-core/libs/log"
+)
+
+// EmbeddedInit defines whenever IPFS repo has to be initialized for an embedded node.
+var EmbeddedInit bool
+
+// Embedded is the provider that embeds IPFS node within the same process.
+// It also returns closable for graceful node shutdown.
+func Embedded(cfg *Config, logger log.Logger) APIProvider {
+	return func() (coreiface.CoreAPI, io.Closer, error) {
+		path := cfg.Path()
+		// NOTE: no need to validate the path before
+		if err := plugins(path); err != nil {
+			return nil, nil, err
+		}
+		// Init Repo if requested
+		if EmbeddedInit {
```

Comment:
This is the only place this global var is read. We should instead just pass it in as a parameter for `Embedded`. The cmd flag should be an unexported var.